File: fitness-monolith/src/main/java/com/project/fitness/util/CalorieCalculator.java
package com.project.fitness.util;

import com.project.fitness.model.ActivityType;

import java.util.Map;

/**
 * Utility for estimating caloric expenditure based on standard Metabolic Equivalent of Task (MET)
 * values defined by the American College of Sports Medicine (ACSM) Compendium of Physical Activities.
 *
 * Formula:
 * Calories = MET * Weight(kg) * (Duration(minutes) / 60) * IntensityMultiplier
 */
public class CalorieCalculator {

    public static final double DEFAULT_WEIGHT_KG = 70.0;

    /**
     * Standard MET values per activity category for moderate intensity.
     */
    public static double getBaseMET(ActivityType type) {
        if (type == null) return 5.0;
        return switch (type) {
            case RUNNING -> 9.8;
            case CYCLING -> 7.5;
            case SWIMMING -> 8.0;
            case WEIGHT_TRAINING -> 5.0;
            case CARDIO -> 8.0;
            case WALKING -> 3.8;
            case YOGA -> 3.0;
            case STRETCHING -> 2.5;
            case OTHER -> 4.5;
        };
    }

    /**
     * Intensity multiplier:
     * - LOW: 0.8x
     * - MODERATE: 1.0x (default)
     * - VIGOROUS / HIGH: 1.25x
     */
    public static double getIntensityMultiplier(String intensity) {
        if (intensity == null) return 1.0;
        return switch (intensity.trim().toUpperCase()) {
            case "LOW" -> 0.8;
            case "VIGOROUS", "HIGH" -> 1.25;
            default -> 1.0;
        };
    }

    /**
     * Calculates estimated calories burned using default fallback weight if none specified.
     */
    public static int calculateCalories(ActivityType type, int durationMinutes, Map<String, Object> additionalMetrics) {
        return calculateCalories(type, durationMinutes, additionalMetrics, null);
    }

    /**
     * Calculates estimated calories burned with user's personal weight fallback.
     *
     * @param type Activity type
     * @param durationMinutes Duration in minutes
     * @param additionalMetrics Optional additional metrics (distance_km, avg_heart_rate, intensity, weight_kg)
     * @param userWeightKg Optional user's profile weight in kg (defaults to DEFAULT_WEIGHT_KG if null)
     * @return Integer estimated calories burned (rounded)
     */
    public static int calculateCalories(ActivityType type, int durationMinutes, Map<String, Object> additionalMetrics, Double userWeightKg) {
        if (durationMinutes <= 0) return 0;

        double weightKg = (userWeightKg != null && userWeightKg > 0) ? userWeightKg : DEFAULT_WEIGHT_KG;
        String intensity = "MODERATE";
        Double distanceKm = null;
        Integer heartRate = null;

        if (additionalMetrics != null) {
            if (additionalMetrics.get("weight_kg") instanceof Number num) {
                weightKg = num.doubleValue();
            }
            if (additionalMetrics.get("intensity") instanceof String s) {
                intensity = s;
            }
            if (additionalMetrics.get("distance_km") instanceof Number num) {
                distanceKm = num.doubleValue();
            }
            if (additionalMetrics.get("avg_heart_rate") instanceof Number num) {
                heartRate = num.intValue();
            }
        }

        // Distance-based refinement for running/cycling if distance is significant
        if (distanceKm != null && distanceKm > 0) {
            if (type == ActivityType.RUNNING) {
                // Running burns ~1.03 kcal per kg per km
                double distanceCalories = distanceKm * weightKg * 1.03;
                return (int) Math.round(distanceCalories);
            } else if (type == ActivityType.WALKING) {
                // Walking burns ~0.53 kcal per kg per km
                double distanceCalories = distanceKm * weightKg * 0.53;
                return (int) Math.round(distanceCalories);
            } else if (type == ActivityType.CYCLING) {
                // Cycling burns ~0.35 - 0.45 kcal per kg per km depending on speed
                double distanceCalories = distanceKm * weightKg * 0.40;
                return (int) Math.round(distanceCalories);
            }
        }

        // Heart rate refinement if available
        double hrMultiplier = 1.0;
        if (heartRate != null && heartRate > 0) {
            if (heartRate >= 160) {
                hrMultiplier = 1.25;
            } else if (heartRate >= 140) {
                hrMultiplier = 1.12;
            } else if (heartRate < 110) {
                hrMultiplier = 0.85;
            }
        }

        double baseMet = getBaseMET(type);
        double intensityMultiplier = getIntensityMultiplier(intensity);

        double calories = baseMet * weightKg * (durationMinutes / 60.0) * intensityMultiplier * hrMultiplier;
        return Math.max(1, (int) Math.round(calories));
    }
}

File: befit-frontend/src/utils/calorieCalculator.js
/**
 * Metabolic Equivalent of Task (MET) values and physiological calorie estimation
 * based on the Compendium of Physical Activities (ACSM).
 *
 * Formula:
 * Calories = MET * Weight(kg) * (Duration(mins) / 60) * IntensityMultiplier * HRMultiplier
 */

export const MET_TABLE = {
  RUNNING: { name: 'Running', met: 9.8, ratePerKm: 1.03, icon: 'Flame' },
  CYCLING: { name: 'Cycling', met: 7.5, ratePerKm: 0.40, icon: 'Bike' },
  SWIMMING: { name: 'Swimming', met: 8.0, ratePerKm: 3.0, icon: 'Waves' },
  WEIGHT_TRAINING: { name: 'Weight Training', met: 5.0, ratePerKm: null, icon: 'Dumbbell' },
  CARDIO: { name: 'HIIT / Cardio', met: 8.0, ratePerKm: null, icon: 'Zap' },
  WALKING: { name: 'Walking', met: 3.8, ratePerKm: 0.53, icon: 'Footprints' },
  YOGA: { name: 'Yoga', met: 3.0, ratePerKm: null, icon: 'Heart' },
  STRETCHING: { name: 'Stretching', met: 2.5, ratePerKm: null, icon: 'Sparkles' },
};

export const INTENSITY_LEVELS = [
  { id: 'LOW', label: 'Low', multiplier: 0.8, description: 'Easy conversational effort' },
  { id: 'MODERATE', label: 'Moderate', multiplier: 1.0, description: 'Elevated breathing, sustained' },
  { id: 'VIGOROUS', label: 'Vigorous', multiplier: 1.25, description: 'Max heart rate, high burn' },
];

export const estimateCalories = ({
  type = 'RUNNING',
  duration = 0,
  intensity = 'MODERATE',
  weightKg = 70,
  distanceKm = null,
  heartRate = null,
}) => {
  const durationMins = Number(duration) || 0;
  if (durationMins <= 0) return 0;

  const activityMeta = MET_TABLE[type] || { met: 5.0 };
  const intensityObj = INTENSITY_LEVELS.find((i) => i.id === intensity) || { multiplier: 1.0 };

  const weight = Number(weightKg) || 70;

  // Distance refinement if provided
  const dist = Number(distanceKm);
  if (dist > 0 && activityMeta.ratePerKm) {
    const distCalories = dist * weight * activityMeta.ratePerKm;
    // Blend distance with duration (80% distance weight, 20% duration intensity)
    return Math.round(distCalories * intensityObj.multiplier);
  }

  // Heart rate refinement
  let hrMultiplier = 1.0;
  const hr = Number(heartRate);
  if (hr > 0) {
    if (hr >= 160) hrMultiplier = 1.25;
    else if (hr >= 140) hrMultiplier = 1.12;
    else if (hr < 110) hrMultiplier = 0.85;
  }

  const baseCalories = activityMeta.met * weight * (durationMins / 60) * intensityObj.multiplier * hrMultiplier;
  return Math.max(1, Math.round(baseCalories));
};
